File: Python/day03/main.py
from collections import defaultdict
from itertools import product


def load_board(filename: str) -> list[(int, list[((int, int), str)])]:
    numbers: list[(int, list[((int, int), str)])] = []
    padded_board: list[str] = []
    with open(filename) as f:
        first_line = f.readline().strip()
        board_width = len(first_line) + 2
        padded_board.append("." * board_width)
        padded_board.append("." + first_line + ".")
        for line in f:
            padded_board.append("." + line.strip() + ".")
        padded_board.append("." * board_width)
    for y, line in enumerate(padded_board):
        number_in_progress = False
        current_number = ""
        start: (int, int) = (0,0)
        for x, char in enumerate(line):
            if char.isdigit():
                if number_in_progress:
                    current_number += char
                else:
                    start = (x - 1, y - 1)
                    number_in_progress = True
                    current_number = char
            else:
                if number_in_progress:
                    number_in_progress = False
                    symbols: list[((int, int), str)] = []
                    for i in range(start[0], x + 1):
                        if padded_board[y - 1][i] != "." and not padded_board[y - 1][i].isdigit():
                            symbols.append(((i, y - 1), padded_board[y - 1][i]))
                    for i in range(start[0], x + 1):
                        if padded_board[y + 1][i] != "." and not padded_board[y + 1][i].isdigit():
                            symbols.append(((i, y + 1), padded_board[y + 1][i]))
                    if padded_board[y][start[0]] != "." and not padded_board[y][start[0]].isdigit():
                        symbols.append(((start[0], y), padded_board[y][start[0]]))
                    if padded_board[y][x] != "." and not padded_board[y][x].isdigit():
                        symbols.append(((x, y), padded_board[y][x]))
                    numbers.append((int(current_number), symbols))
    return numbers


def solution_part_1():
    board = load_board("input.txt")
    part_number_sum = 0
    for number in board:
        value, symbols = number
        if len(symbols) > 0:
            part_number_sum += value
    print(part_number_sum)


def solution_part_2():
    board = load_board("input.txt")
    numbers_for_stars: dict[(int, int), list[int]] = defaultdict(lambda: [])
    gear_ratio_sum = 0
    for number in board:
        value, symbols = number
        for symbol in symbols:
            if symbol[1] == "*":
                numbers_for_stars[symbol[0]].append(value)
    for star in numbers_for_stars:
        if len(numbers_for_stars[star]) == 2:
            gear_ratio_sum += numbers_for_stars[star][0] * numbers_for_stars[star][1]
    print(gear_ratio_sum)


def main():
    solution_part_1()
    solution_part_2()


if __name__ == "__main__":
    main()

File: golang/day03/main.go
package main

import (
	"bufio"
	"os"
	"strconv"
	"unicode"
)

type Point struct {
	x int
	y int
}
type Symbol struct {
	value    rune
	location Point
}

type Number struct {
	Value   int
	Symbols []Symbol
}

func padLine(line []rune, padding rune) []rune {
	return append(append([]rune{padding}, line...), padding)
}

func filledLine(width int, fill rune) []rune {
	line := make([]rune, width)
	for i := 0; i < width; i++ {
		line[i] = fill
	}
	return line
}

func loadBoard(filename string) ([]Number, error) {
	data, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer func(data *os.File) {
		err := data.Close()
		if err != nil {
			panic(err)
		}
	}(data)
	scanner := bufio.NewScanner(data)
	numbers := make([]Number, 0)
	paddedBoard := make([][]rune, 0)
	// get first line to get width
	scanner.Scan()
	firstLine := scanner.Text()
	width := len(firstLine)
	paddedBoard = append(paddedBoard, filledLine(width, '.'))
	paddedBoard = append(paddedBoard, padLine([]rune(firstLine), '.'))
	for scanner.Scan() {
		line := scanner.Text()
		paddedBoard = append(paddedBoard, padLine([]rune(line), '.'))
	}
	paddedBoard = append(paddedBoard, filledLine(width, '.'))
	if err := scanner.Err(); err != nil {
		return nil, err
	}
	// find numbers
	for y, line := range paddedBoard {
		numberInProgress := false
		var currentNumber []rune
		var start Point
		for x, char := range line {
			if unicode.IsDigit(char) {
				if numberInProgress {
					currentNumber = append(currentNumber, char)
				} else {
					numberInProgress = true
					currentNumber = make([]rune, 0)
					currentNumber = append(currentNumber, char)
					start = Point{x: x - 1, y: y - 1}
				}
			} else {
				if numberInProgress {
					parsedNumber, err := strconv.Atoi(string(currentNumber))
					if err != nil {
						return nil, err
					}
					numbers = append(numbers, Number{Value: parsedNumber, Symbols: getSymbols(paddedBoard, start, Point{x, y + 1})})
					numberInProgress = false
				}
			}
		}
	}
	return numbers, nil
}

func getSymbols(board [][]rune, start Point, end Point) []Symbol {
	symbols := make([]Symbol, 0)
	for i := start.x; i <= end.x; i++ {
		if board[start.y][i] != '.' && !unicode.IsDigit(board[start.y][i]) {
			symbols = append(symbols, Symbol{value: board[start.y][i], location: Point{x: i, y: start.y}})
		}
		if board[end.y][i] != '.' && !unicode.IsDigit(board[end.y][i]) {
			symbols = append(symbols, Symbol{value: board[end.y][i], location: Point{x: i, y: end.y}})
		}
	}
	if board[start.y+1][start.x] != '.' && !unicode.IsDigit(board[start.y+1][start.x]) {
		symbols = append(symbols, Symbol{value: board[start.y+1][start.x], location: Point{x: start.x, y: start.y + 1}})
	}
	if board[start.y+1][end.x] != '.' && !unicode.IsDigit(board[start.y+1][end.x]) {
		symbols = append(symbols, Symbol{value: board[start.y+1][end.x], location: Point{x: end.x, y: start.y + 1}})
	}
	return symbols
}

func solutionPart1() {
	board, err := loadBoard("input.txt")
	if err != nil {
		panic(err)
	}
	partNumbersSum := 0
	for _, number := range board {
		if len(number.Symbols) > 0 {
			partNumbersSum += number.Value
		}
	}
	println(partNumbersSum)
}

func solutionPart2() {
	board, err := loadBoard("input.txt")
	if err != nil {
		panic(err)
	}
	gearNumbers := make(map[Point][]int)
	for _, number := range board {
		for _, symbol := range number.Symbols {
			if symbol.value == '*' {
				gearNumbers[symbol.location] = append(gearNumbers[symbol.location], number.Value)
				break
			}
		}
	}
	gearRatioSum := 0
	for _, gearNumber := range gearNumbers {
		if len(gearNumber) == 2 {
			gearRatioSum += gearNumber[0] * gearNumber[1]
		}
	}
	println(gearRatioSum)
}

func main() {
	solutionPart1()
	solutionPart2()
}
